api/pkg/s3/objectget.go

```
@@ -24,65 +33,152 @@ import (
+			response.ResponseWriter.Header()["ETag"] = []string{"\"" + object.Etag + "\""}
+		}
+		if err == s3error.ContentNotModified { // write only header if is a 304
+			WriteErrorResponseHeaders(response, err)
+		} else {
+			WriteErrorResponse(response, request, err)
+		}
 		return
 	}
-	log.Infof("enter the download method")
-	body, s3err := client.GET(&object, ctx, int64(start), int64(end))
-	log.Infof("out  the download method")
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+
+	// Get the object.
+	startOffset := int64(0)
+	length := object.Size
+	if hrange != nil {
+		startOffset = hrange.OffsetBegin
+		length = hrange.GetLength()
+	}
+	stream, err := s.s3Client.GetObject(ctx, &pb.GetObjectInput{Bucket:bucketName, Key:objectKey, Offset: startOffset, Length: length,})
+	if err != nil {
+		log.Errorln("get object failed, err:%v", err)
+		WriteErrorResponse(response, request, err)
 		return
 	}
-	buf := new(bytes.Buffer)
-	buf.ReadFrom(body)
-	response.Write(buf.Bytes())
-	*/
-	log.Infof("Init multipart upload[bucketName=%s, objectKey=%s] successfully.\n", bucketName, objectKey)
+	defer stream.Close()
+
+	// Indicates if any data was written to the http.ResponseWriter
+	dataWritten := false
+	// io.Writer type which keeps track if any data was written.
+	writer := funcToWriter(func(p []byte) (int, error) {
+		if !dataWritten {
+			// Set headers on the first write.
+			// Set standard object headers.
+			SetObjectHeaders(response, object, hrange)
+
+			// Set any additional requested response headers.
+			setGetRespHeaders(response.ResponseWriter, request.Request.URL.Query())
+			dataWritten = true
+		}
+		n, err := response.Write(p)
+		if n > 0 {
+
+		}
+		return n, err
+	})
+
+	s3err := int32(0)
+	eof := false
+	left := length
+	for !eof && left > 0 {
+		rsp, err := stream.Recv()
+		if err != nil && err != io.EOF {
+			log.Errorln("recv err", err)
+			break
+		}
+		if err == io.EOF {
+			eof = true
```

it is good to break directly here.

---

A non-zero number of bytes may be returned with error io.eof.

---

Does it seem the error of io.eof is not returned anywhere?

---

1. if all data have been sent to end user, we can ignore the error io.eof
2. if partial data have been sent,  the content-length is not equal to sent data length, and the request is  failed